app/javascript/pages/map/recent-imagery/recent-imagery-actions.js

```diff
@@ -1,38 +1,94 @@
 import { createAction } from 'redux-actions';
 import { createThunkAction } from 'utils/redux';
+import { CancelToken } from 'axios';
+import findIndex from 'lodash/findIndex';
 
-import { getRecentTiles } from 'services/recent-imagery';
+import { getRecentTiles, getTiles, getThumbs } from 'services/recent-imagery';
 
 const setRecentImageryData = createAction('setRecentImageryData');
 const toogleRecentImagery = createAction('toogleRecentImagery');
-const setRecentImagerySettingsStyles = createAction(
-  'setRecentImagerySettingsStyles'
+const setRecentImagerySettings = createAction('setRecentImagerySettings');
+const setRecentImageryShowSettings = createAction(
+  'setRecentImageryShowSettings'
 );
 
-const getTiles = createThunkAction('getTiles', params => dispatch => {
-  getRecentTiles(params)
+const getData = createThunkAction('getData', params => dispatch => {
+  if (this.getDataSource) {
+    this.getDataSource.cancel();
+  }
+  this.getDataSource = CancelToken.source();
+
+  getRecentTiles({ ...params, token: this.getDataSource.token })
     .then(response => {
-      if (response.data.data.length) {
-        const data = response.data.data[0].attributes;
-        dispatch(
-          setRecentImageryData({
-            url: data.tile_url,
-            bounds: data.bbox.geometry.coordinates,
-            cloudScore: data.cloud_score,
-            dateTime: data.date_time,
-            instrument: data.instrument
-          })
-        );
+      if (response.data.data.tiles) {
+        dispatch(setRecentImageryData({ data: response.data.data }));
+        dispatch(setRecentImagerySettings({ selectedTileIndex: 0 }));
       }
     })
     .catch(error => {
       console.error(error);
     });
 });
 
+const getMoreTiles = createThunkAction(
+  'getMoreTiles',
+  params => (dispatch, state) => {
+    if (this.getMoreTilesSource) {
+      this.getMoreTilesSource.cancel();
+    }
+    this.getMoreTilesSource = CancelToken.source();
+
+    getTiles({ ...params, token: this.getMoreTilesSource.token })
+      .then(getTilesResponse => {
+        getThumbs({ ...params, token: this.getMoreTilesSource.token })
+          .then(getThumbsResponse => {
+            if (
+              getTilesResponse.data.data &&
+              getTilesResponse.data.data.attributes.length &&
+              getThumbsResponse.data.data &&
+              getThumbsResponse.data.data.attributes.length
+            ) {
+              const stateData = state().recentImagery.data;
+              const data = { ...stateData, tiles: stateData.tiles.slice() };
+              getTilesResponse.data.data.attributes.forEach((item, i) => {
```

Comment:
This is very hard to follow and a little scary 😨 . Could you assign nested data when accessing nested elements or array indexs inside object. It is hard to know if you will have an undefined in there.